File: picasso-pollexor/src/test/java/com/squareup/picasso/pollexor/PollexorRequestTransformerTest.java
package com.squareup.picasso.pollexor;

import android.net.Uri;
import com.squareup.picasso.Request;
import com.squareup.pollexor.Thumbor;
import com.squareup.pollexor.ThumborUrlBuilder.ImageFormat;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricGradleTestRunner;
import org.robolectric.annotation.Config;

import static com.google.common.truth.Truth.assertThat;
import static com.squareup.picasso.Picasso.RequestTransformer;
import static com.squareup.pollexor.ThumborUrlBuilder.format;
import static org.robolectric.annotation.Config.NONE;

@RunWith(RobolectricGradleTestRunner.class)
@Config(
    sdk = 17,
    manifest = NONE,
    constants = BuildConfig.class
)
public class PollexorRequestTransformerTest {
  private static final String HOST = "http://example.com/";
  private static final String KEY = "omgsecretpassword";
  private static final String IMAGE = "http://google.com/logo.png";
  private static final Uri IMAGE_URI = Uri.parse(IMAGE);

  private RequestTransformer transformer = new PollexorRequestTransformer(Thumbor.create(HOST));
  private RequestTransformer secureTransformer =
      new PollexorRequestTransformer(Thumbor.create(HOST, KEY));

  @Test public void resourceIdRequestsAreNotTransformed() {
    Request input = new Request.Builder(12).build();
    Request output = transformer.transformRequest(input);
    assertThat(output).isSameAs(input);
  }

  @Test public void nonHttpRequestsAreNotTransformed() {
    Request input = new Request.Builder(IMAGE_URI).build();
    Request output = transformer.transformRequest(input);
    assertThat(output).isSameAs(input);
  }

  @Test public void nonResizedRequestsAreNotTransformed() {
    Request input = new Request.Builder(IMAGE_URI).build();
    Request output = transformer.transformRequest(input);
    assertThat(output).isSameAs(input);
  }

  @Test public void simpleResize() {
    Request input = new Request.Builder(IMAGE_URI).resize(50, 50).build();
    Request output = transformer.transformRequest(input);
    assertThat(output).isNotSameAs(input);
    assertThat(output.hasSize()).isFalse();

    String expected = Thumbor.create(HOST).buildImage(IMAGE).resize(50, 50).toUrl();
    assertThat(output.uri.toString()).isEqualTo(expected);
  }

  @Config(sdk = 18)
  @Test public void simpleResizeOnJbMr2UsesWebP() {
    Request input = new Request.Builder(IMAGE_URI).resize(50, 50).build();
    Request output = transformer.transformRequest(input);
    assertThat(output).isNotSameAs(input);
    assertThat(output.hasSize()).isFalse();

    String expected = Thumbor.create(HOST)
        .buildImage(IMAGE)
        .resize(50, 50)
        .filter(format(ImageFormat.WEBP))
        .toUrl();
    assertThat(output.uri.toString()).isEqualTo(expected);
  }

  @Test public void simpleResizeWithCenterCrop() {
    Request input = new Request.Builder(IMAGE_URI).resize(50, 50).centerCrop().build();
    Request output = transformer.transformRequest(input);
    assertThat(output).isNotSameAs(input);
    assertThat(output.hasSize()).isFalse();
    assertThat(output.centerCrop).isFalse();

    String expected = Thumbor.create(HOST).buildImage(IMAGE).resize(50, 50).toUrl();
    assertThat(output.uri.toString()).isEqualTo(expected);
  }

  @Test public void simpleResizeWithCenterInside() {
    Request input = new Request.Builder(IMAGE_URI).resize(50, 50).centerInside().build();
    Request output = transformer.transformRequest(input);
    assertThat(output).isNotSameAs(input);
    assertThat(output.hasSize()).isFalse();
    assertThat(output.centerInside).isFalse();

    String expected = Thumbor.create(HOST).buildImage(IMAGE).resize(50, 50).fitIn().toUrl();
    assertThat(output.uri.toString()).isEqualTo(expected);
  }

  @Test public void simpleResizeWithEncryption() {
    Request input = new Request.Builder(IMAGE_URI).resize(50, 50).build();
    Request output = secureTransformer.transformRequest(input);
    assertThat(output).isNotSameAs(input);
    assertThat(output.hasSize()).isFalse();

    String expected = Thumbor.create(HOST, KEY).buildImage(IMAGE).resize(50, 50).toUrl();
    assertThat(output.uri.toString()).isEqualTo(expected);
  }

  @Test public void simpleResizeWithCenterInsideAndEncryption() {
    Request input = new Request.Builder(IMAGE_URI).resize(50, 50).centerInside().build();
    Request output = secureTransformer.transformRequest(input);
    assertThat(output).isNotSameAs(input);
    assertThat(output.hasSize()).isFalse();
    assertThat(output.centerInside).isFalse();

    String expected = Thumbor.create(HOST, KEY).buildImage(IMAGE).resize(50, 50).fitIn().toUrl();
    assertThat(output.uri.toString()).isEqualTo(expected);
  }
}

File: picasso-pollexor/src/test/java/com/squareup/picasso3/pollexor/PollexorRequestTransformerTest.kt
/*
 * Copyright (C) 2022 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.squareup.picasso3.pollexor

import android.net.Uri
import com.google.common.truth.Truth.assertThat
import com.squareup.picasso3.Request.Builder
import com.squareup.pollexor.Thumbor
import com.squareup.pollexor.ThumborUrlBuilder
import com.squareup.pollexor.ThumborUrlBuilder.ImageFormat.WEBP
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class PollexorRequestTransformerTest {
  private val transformer = PollexorRequestTransformer(Thumbor.create(HOST))
  private val secureTransformer = PollexorRequestTransformer(Thumbor.create(HOST, KEY))
  private val alwaysResizeTransformer = PollexorRequestTransformer(
    Thumbor.create(HOST), alwaysTransform = true
  )
  private val callbackTransformer = PollexorRequestTransformer(
    Thumbor.create(HOST), callback = { it.filter("custom") }
  )

  @Test fun resourceIdRequestsAreNotTransformed() {
    val input = Builder(12).build()
    val output = transformer.transformRequest(input)
    assertThat(output).isSameInstanceAs(input)
  }

  @Test fun resourceIdRequestsAreNotTransformedWhenAlwaysTransformIsTrue() {
    val input = Builder(12).build()
    val output = alwaysResizeTransformer.transformRequest(input)
    assertThat(output).isSameInstanceAs(input)
  }

  @Test fun nonHttpRequestsAreNotTransformed() {
    val input = Builder(IMAGE_URI).build()
    val output = transformer.transformRequest(input)
    assertThat(output).isSameInstanceAs(input)
  }

  @Test fun nonResizedRequestsAreNotTransformed() {
    val input = Builder(IMAGE_URI).build()
    val output = transformer.transformRequest(input)
    assertThat(output).isSameInstanceAs(input)
  }

  @Test fun nonResizedRequestsAreTransformedWhenAlwaysTransformIsSet() {
    val input = Builder(IMAGE_URI).build()
    val output = alwaysResizeTransformer.transformRequest(input)
    assertThat(output).isNotSameInstanceAs(input)
    assertThat(output.hasSize()).isFalse()

    val expected = Thumbor.create(HOST)
      .buildImage(IMAGE)
      .filter(ThumborUrlBuilder.format(WEBP))
      .toUrl()
    assertThat(output.uri.toString()).isEqualTo(expected)
  }

  @Test fun simpleResize() {
    val input = Builder(IMAGE_URI).resize(50, 50).build()
    val output = transformer.transformRequest(input)
    assertThat(output).isNotSameInstanceAs(input)
    assertThat(output.hasSize()).isFalse()

    val expected = Thumbor.create(HOST)
      .buildImage(IMAGE)
      .resize(50, 50)
      .filter(ThumborUrlBuilder.format(WEBP))
      .toUrl()
    assertThat(output.uri.toString()).isEqualTo(expected)
  }

  @Test fun simpleResizeWithCenterCrop() {
    val input = Builder(IMAGE_URI).resize(50, 50).centerCrop().build()
    val output = transformer.transformRequest(input)
    assertThat(output).isNotSameInstanceAs(input)
    assertThat(output.hasSize()).isFalse()
    assertThat(output.centerCrop).isFalse()

    val expected = Thumbor.create(HOST)
      .buildImage(IMAGE)
      .resize(50, 50)
      .filter(ThumborUrlBuilder.format(WEBP))
      .toUrl()
    assertThat(output.uri.toString()).isEqualTo(expected)
  }

  @Test fun simpleResizeWithCenterInside() {
    val input = Builder(IMAGE_URI).resize(50, 50).centerInside().build()
    val output = transformer.transformRequest(input)
    assertThat(output).isNotSameInstanceAs(input)
    assertThat(output.hasSize()).isFalse()
    assertThat(output.centerInside).isFalse()

    val expected = Thumbor.create(HOST)
      .buildImage(IMAGE)
      .resize(50, 50)
      .filter(ThumborUrlBuilder.format(WEBP))
      .fitIn()
      .toUrl()
    assertThat(output.uri.toString()).isEqualTo(expected)
  }

  @Test fun simpleResizeWithEncryption() {
    val input = Builder(IMAGE_URI).resize(50, 50).build()
    val output = secureTransformer.transformRequest(input)
    assertThat(output).isNotSameInstanceAs(input)
    assertThat(output.hasSize()).isFalse()

    val expected = Thumbor.create(HOST, KEY)
      .buildImage(IMAGE)
      .resize(50, 50)
      .filter(ThumborUrlBuilder.format(WEBP))
      .toUrl()
    assertThat(output.uri.toString()).isEqualTo(expected)
  }

  @Test fun simpleResizeWithCenterInsideAndEncryption() {
    val input = Builder(IMAGE_URI).resize(50, 50).centerInside().build()
    val output = secureTransformer.transformRequest(input)
    assertThat(output).isNotSameInstanceAs(input)
    assertThat(output.hasSize()).isFalse()
    assertThat(output.centerInside).isFalse()

    val expected = Thumbor.create(HOST, KEY)
      .buildImage(IMAGE)
      .resize(50, 50)
      .filter(ThumborUrlBuilder.format(WEBP))
      .fitIn()
      .toUrl()
    assertThat(output.uri.toString()).isEqualTo(expected)
  }

  @Test fun configureCallback() {
    val input = Builder(IMAGE_URI).resize(50, 50).build()
    val output = callbackTransformer.transformRequest(input)
    assertThat(output).isNotSameInstanceAs(input)
    assertThat(output.hasSize()).isFalse()
    val expected = Thumbor.create(HOST)
      .buildImage(IMAGE)
      .resize(50, 50)
      .filter("custom")
      .filter(ThumborUrlBuilder.format(WEBP))
      .toUrl()
    assertThat(output.uri.toString()).isEqualTo(expected)
  }

  companion object {
    private const val HOST = "http://example.com/"
    private const val KEY = "omgsecretpassword"
    private const val IMAGE = "http://google.com/logo.png"
    private val IMAGE_URI = Uri.parse(IMAGE)
  }
}
